### ISSUE TITLE
Project dependencies may have API risk issues

### ISSUE THREAD

**Issue by PyDR-Tool** [facebookresearch/pifuhd#186]
Hi, In **pifuhd**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using
```
torchvision
Pillow
tqdm
trimesh
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **tqdm** can be changed to *>=4.6.1,<=4.64.0*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

In version **tqdm-4.5.2**, the API **tqdm.tqdm_notebook** whch is used by the current project in **lib/colab_util.py* is missing.

<img width="381" alt="image" src="https://user-images.githubusercontent.com/109138844/227120658-267f59fb-3100-435a-a304-d3a9ba7d0897.png">

<details>
<summary>The calling methods from the tqdm</summary>
<pre>
tqdm.tqdm_notebook
tqdm.tqdm
</pre>
</details>

    
<details>
<summary>The calling methods from the  all methods</summary>
<pre>
self._forward
points.clone
self.get_translation_vector
pts.center.sum
images.size.nn.Upsample.append
argparse.ArgumentParser.add_argument
load_obj_mesh
net.mlp.state_dict
argparse.ArgumentParser.parse_args
calib_local.size
lib.mesh_util.reconstruction
bbox.max
net.filter_global
nmls.torch.stack.view
glCreateShader
glVertexAttribPointer
torch.cat
open.close
intrinsic.torch.Tensor.float
glGenTextures
torch.cuda.is_available
total_vals.append
numpy.clip
glCompileShader
self.conv2
MeshEvaluator.get_chamfer_dist
self.spatial_enc
self.gather_options
MeshEvaluator.get_reproj_normal_error
self.children
cv2.decomposeProjectionMatrix
PIL.Image.fromarray
x.split
l_face_data.append
mesh_file.split
MeshEvaluator
numpy.ones_like
cv2.resize.astype
self.im_feat_list.view.view
self.build_conv_block
outputs.append
Encoder.cuda
cv2.imread
torch.nn.init.normal_
items.append
numpy.sin
self.CustomMSELoss.super.__init__
IOError
output_ins.torch.mean.expand_as
torch.nn.Dropout
calib_tensor.cpu.numpy
torch.nn.ReLU
self.bn2
int
norm_layer
obj_path.split
numpy.linalg.det
glutCreateWindow
torch.stack
torch.no_grad
pred.view.view
Encoder
functools.partial
pytorch3d.renderer.PointLights
numpy.logical_and
points.size
uv.unsqueeze.unsqueeze
torch.nn.DataParallel.load_state_dict
net.nmls.detach.cpu.numpy
pts.max
model_upsample
intrinsic.torch.Tensor.float.unsqueeze
loadShader
i.self.norms
glBindRenderbuffer
glTexImage2DMultisample
center.ps.reshape.sum
list.read
rgb_colors.append
torch.nn.init.orthogonal_
self.draw
self.mlp.state_dict.items
dot
glTexImage2D
self.HGPIFuMRNet.super.__init__
LocalEnhancer
model_downsample
self.camera.get_gl_matrix
torchvision.transforms.ToTensor
cv2.imwrite
self._render_normal
self.HGFilter.super.__init__
glDrawArrays
numpy.clip.cpu
str
numpy.concatenate.astype
self._normal_render.draw
self.filter_global
vals.append
self.error_term
inst.cpu.numpy.astype
render.Render.__init__
self.set_rotation_matrix
glDepthFunc
open.readlines
sdf.eval_grid_octree
torch.nn.functional.leaky_relu
glutInitDisplayMode
torch.log
newlabels.size
torch.mean
glUniformMatrix4fv
reconWrapper
net_util.init_net
json.load
self.sanity_check
pytorch3d.renderer.OpenGLOrthographicCameras
pred.detach.cpu
torch.nn.AvgPool2d
preds_low.append
DepthNormalizer.DepthNormalizer
glAttachShader
self.bn3
in_bb.size.in_bb.view.sum
abs
numpy.expand_dims
bbox.min
newlabels.size.newlabels.view.sum
os.makedirs
getattr
self.direction.reshape
ConvBlock
lib.options.BaseOptions.parse
cv2.VideoWriter_fourcc
numpy.save
glDrawBuffer
init_weights
numpy.sqrt
lib.options.BaseOptions
data.to
face_uv_data.append
argparse.ArgumentParser
torch.nn.ConvTranspose2d
glGenBuffers
self._get_reproj_normal_error
preds.append
torch.nn.functional.relu
sdf.eval_grid
self.criteria
newlabels.view
self.crop_func
self.DepthNormalizer.super.__init__
self.color_buffer.append
glm.ortho
torch.nn.Conv1d
torch.nn.init.constant_
torch.nn.Linear
self.norm_buffer.keys
super
faces_clr.reshape
m.weight.data.normal_
images_w_tex.cpu.numpy
numpy.isnan
glGetShaderiv
range
glCreateProgram
torch.Tensor
lib.render.gl.geo_render.GeoRender
gt.pred.pow.weight.mean
lib.model.HGPIFuNetwNML
parser.add_argument_group.add_argument
os.path.basename
self.parser.get_default
projection_mode.opt_netG.HGPIFuNetwNML.to.eval
v.image_tensor_global.detach.cpu.numpy
torchvision.transforms.Normalize
lib.render.gl.color_render.ColorRender.get_color
set_eval
NameError
hasattr
self.to_tensor.unsqueeze
list
face_data_mat.keys
faces.reshape
pmin.pmax.astype
compute_normal
attachments.append
self.bn1
glGenRenderbuffers
torch.nn.Upsample
self.projection_matrix.transpose
numpy.matmul
lib.geometry.index
glutInitWindowPosition
self.projection
numpy.where
self.model.clone
v.image_tensor_global.detach.cpu
os.path.abspath
torch.nn.Sequential
face_data.append
glEnable
glClear
obj_files.append
pid.rects.tolist
im.Image.fromarray.convert
i.str.self._modules
total_vals.np.array.mean
pred.view.size
coords.T.reshape
setattr
self.preds_low.size
line.split.split
normalize
l_face_norm_data.append
self.get_rotation_matrix
pytorch3d.renderer.look_at_view_transform
torch.nn.DataParallel.apply
self.to_tensor
self.LocalEnhancer.super.__init__
sorted
k.model_dict.size
torch.nn.DataParallel.to
self.conv1
torch.nn.functional.normalize.size
points_sub.size
set.add
in_bb.detach.float.view
magnitude
data.strip.split.split
preds_interm.append
PIL.Image.fromarray.save
obj_path.replace
numpy.flip
w.loss.mean
k.pretrained_dict.size
i.int.b.b.inst.nonzero
self.ResnetBlock.super.__init__
glutMainLoop
glActiveTexture
in_bb.detach.float.detach
net.get_preds
glutDisplayFunc
net_util.conv3x3
net.mlp.state_dict.items
intermediate_preds_list.append
self.HourGlass.super.__init__
glBindTexture
isinstance
center.ps.reshape
glUniform1i
glutPostRedisplay
torch.from_numpy
f.read
glutSwapBuffers
gt.pred.pow
open
self.down_conv2
torch.nn.Tanh
self.MLP.super.__init__
identity
numpy.eye
demo_list.append
torch.nn.functional.avg_pool2d
pytorch3d.io.load_objs_as_meshes
projection_matrix.torch.Tensor.float.unsqueeze
self.filter
self.downsample
points.torch.from_numpy.to
self.screen_texture.append
renderer
net.filter_local
torch.nn.BatchNorm2d
self.opt.loadSizeBig.self.opt.loadSizeBig.nn.Upsample.expand
numpy.stack.mean
self.im_feat_list.view
torch.nn.Sigmoid
HGFilters.HGFilter
numpy.radians
cv2.cvtColor
data.strip.split
projection_mode.netG.opt.HGPIFuMRNet.to.load_state_dict
numpy.transpose
math.sin
net.nmls.detach.cpu
uv.image_tensor.index.detach
torch.cat.size
self.mlp.state_dict
torch.nn.ELU
self._normal_render.set_normal_mesh
glClampColor
createProgram
lib.render.camera.Camera
pytorch3d.structures.Meshes
self.add_module
self.last_op.clone
torch.nn.ModuleList
face_norm_data_mat.keys
verts.T.torch.from_numpy.unsqueeze.to
lib.model.HGPIFuMRNet
set
torch.nn.functional.normalize
numpy.arange
gammas.append
pytorch3d.renderer.MeshRasterizer
self.netF.forward
cam_render.CamRender.__init__
numpy.stack.max
pytorch3d.renderer.MeshRenderer
inst.cpu.numpy
glDrawBuffers
self.model
torch.nn.functional.interpolate
line.decode.strip
self._generate_network
uv.image_tensor.index.detach.cpu.numpy
param.numel
trimesh.sample.sample_surface
glBlitFramebuffer
tgt_src_dist.mean.mean
numpy.linalg.norm
images.self.netF.forward.detach
index.self.img_files.split
make_rotate
gen_mesh
torch.nn.SELU
self.im_feat_list.size
net.mlp.state_dict.items.items
in_bb.detach.float
self.Encoder.super.__init__
vertex_data.append
torch.nn.init.xavier_normal_
uv.unsqueeze.transpose
trimesh.load.split
pred.detach.cpu.numpy
glGetProgramInfoLog
self.GlobalGenerator.super.__init__
uv.image_tensor.index.detach.cpu
MeshEvaluator.set_mesh
self.CustomBCELoss.super.__init__
lib.render.gl.color_render.ColorRender.set_camera
self.opt.loadSizeBig.self.opt.loadSizeBig.nn.Upsample
glShaderSource
net.parameters
empty
min
selected_data.np.array.reshape
norm_data.append
glLinkProgram
self.initialize.add_argument
HourGlass
self.BasePIFuNet.super.__init__
data.strip
data.numpy
v.size
self.normalize_vector
torch.nn.ReflectionPad2d
projection_mode.opt_netG.HGPIFuNetwNML.to
net.calc_normal
mp4.b64encode.decode
tgt_normal.src_normal.mean
vars
f.read.splitlines
meshcleaning
numpy.concatenate
self.image_filter.load_state_dict
self.netB.forward
self.image_filter.state_dict
numpy.stack
self.initialize
net.state_dict.items
glUseProgram
cv2.copyMakeBorder
lib.render.gl.color_render.ColorRender
self.conv4
run
gt.pred.pow.weight.mean.mean
tmp.np.array.reshape
self.mlp
faces_nml.reshape
sdf.create_grid
math.cos
save_img_list.append
torchvision.transforms.Compose
cross
self.opt.loadSizeBig.self.opt.loadSizeBig.nn.Upsample.append
vertices.min
self.right.reshape
open.write
numpy.identity
classname.find
networks.define_G
normalize_v3
line.split
state_dict.items
rect.size
cv2.resize
glutInit
points_sub.clone
MeshEvaluator.get_surface_dist
self.draw_end
zip
w.size
images.self.netB.forward.detach
self.filters.append
self._normal_render.get_color
self.get_extrinsic_matrix
self.last_op
lib.render.mesh.compute_normal
glDetachShader
Exception
labels.view
glGetShaderInfoLog
numpy.repeat
self.netG.filter
pmin.pmax.astype.astype
enumerate
nml.torch.stack.view
glGetProgramiv
glReadPixels
trimesh.load
numpy.ones
float
eval_func
numpy.zeros
glBindFramebuffer
self.query
os.system
im_512.Image.fromarray.convert
numpy.linalg.inv
lib.render.gl.color_render.ColorRender.set_mesh
self.vert_buffer.keys
glReadBuffer
numpy.copy
trimesh.proximity.closest_point
module.train
self.set_intrinsic_matrix
glFramebufferRenderbuffer
os.path.exists
f
pytorch3d.renderer.RasterizationSettings
numpy.meshgrid
self.conv_block
get_norm_layer
Encoder.apply
pytorch3d.renderer.TexturesVertex
MeshEvaluator.init_gl
torch.nn.BatchNorm1d
self.mlp.load_state_dict
labels.size
projection_mode.netG.opt.HGPIFuMRNet.to
f.split
recon
glBufferData
self.up.reshape
in_bb.detach.float.size
hgnet.state_dict
glutKeyboardFunc
glFramebufferTexture2D
z_feat.detach.detach
glDisable
skimage.measure.marching_cubes_lewiner
glDeleteBuffers
tqdm.tqdm
len
coords.T.reshape.reshape
self.calc_normal
torch.nn.ReplicationPad2d
glDisableVertexAttribArray
NotImplementedError
torch.nn.utils.weight_norm
pts.center.sum.np.sqrt.max
numpy.dot
src_tgt_dist.mean.mean
sum
pytorch3d.renderer.HardPhongShader
euler_to_rot_mat
base64.b64encode
f.replace
numpy.sum
shader_list.append
ws.append
numpy.stack.append
crop_image
glGenFramebuffers
c_void_p
projection_matrix.torch.Tensor.float
torch.nn.DataParallel
GlobalGenerator
torch.nn.LeakyReLU
line.decode.split
face_normals.reshape
input.size
uv_data.append
m.bias.data.fill_
self.preds_interm.size
mesh_file.replace
self.ConvBlock.super.__init__
hgnet.state_dict.items
numpy.unique
read_mtlfile
torch.nn.DataParallel.state_dict
max
glTexParameteri
numpy.max
numpy.loadtxt
os.path.dirname
self.get_intrinsic_matrix
face_norm_data.append
net.get_preds.detach
numpy.tan
lib.data.EvalDataset
render.gl.normal_render.NormalRender
torch.nn.functional.normalize.append
lib.mesh_util.save_obj_mesh_with_color
line.decode.startswith
input_downsampled.append
torch.baddbmm
calib_tensor.cpu
batch_eval
verts.T.torch.from_numpy.unsqueeze
torch.load
self.netG.query
self.get_preds
inst.cpu
self.filter_local
os.path.splitext
torch.nn.GroupNorm
get_verts_rgb_colors
tuple
self.draw_init
camera.get_gl_matrix
torch.nn.init.kaiming_normal_
self.get_error
self.model_view_matrix.transpose
face_uv_data_mat.keys
numpy.zeros.reshape
line.decode.decode
self.image_filter
images.size.nn.Upsample
torch.nn.MSELoss
os.path.join.split
pts.astype
self.init_quad_program
self._normal_render.set_matrices
line.strip.split
self.center.reshape
self.parser.parse_args
state_dict.items.items
Rot.dot
self.HGPIFuNetwNML.super.__init__
self.index
sys.path.insert
vertices.max
os.path.join
net.nmls.detach
net.projection
cv2.VideoWriter.release
k.split
cuda.points.torch.from_numpy.to.float
lib.data.EvalWPoseDataset.get_n_person
cuda.verts.T.torch.from_numpy.unsqueeze.to.float
path.open.read
net.query
numpy.cos
level.str.self._modules
glReadPixels.reshape
torch.nn.Conv2d
l_face_uv_data.append
os.path.isfile
self.get_item
glClearColor
verts_rgb_colors.torch.from_numpy.to
glBindBuffer
cam_render.CamRender.set_mesh
torch.nn.functional.grid_sample
points_all.view.view
newlabels.append
self.color_buffer.keys
torch.cuda.set_device
torch.device
pts.min
glGetUniformLocation
self.netG.eval
numpy.cross
glEnableVertexAttribArray
format
cv2.VideoWriter.write
torch.cat.view
torch.clamp
out_mesh.export
vec3
cv2.VideoWriter
ResnetBlock
glRenderbufferStorageMultisample
index.self.img_files.replace
opt.vars.items
self.initialize.add_argument_group
total_vals.np.array.mean.append
MLP.MLP
tqdm.tqdm_notebook
v.image_tensor_global.detach
numpy.array
os.walk
lib.render.gl.color_render.ColorRender.display
lib.data.EvalWPoseDataset
glDeleteShader
math.radians
print
findFileOrThrow
IPython.display.HTML
coords.reshape.reshape
numpy.stack.min
torch.nn.PReLU
self.conv3
images.size.nn.Upsample.size
glutInitWindowSize
os.listdir
self.im_feat_list.view.size
camera_center.reshape.reshape
glClearDepth
self.norms.append
compute_tangent
map
torch.nn.InstanceNorm1d
</pre>
</details>

    

@developer
Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.
